onnxruntime/core/providers/cpu/generator/constant_of_shape_base.h

```diff
@@ -0,0 +1,135 @@
+        p_value_ = reinterpret_cast<void*>(&(s_value_.int32_));
+        break;
+      case sizeof(int64_t):
+        s_value_.int64_ = *(reinterpret_cast<int64_t*>(value));
+        p_value_ = reinterpret_cast<void*>(&(s_value_.int64_));
+        break;
+      default:
+        ORT_THROW("Unsupported value attribute datatype with size: ", size);
+    }
+  }
+
+  void SetValueFromTensorProto(const ONNX_NAMESPACE::TensorProto&);
+};
+
+#define CASE_FETCH_VALUE_DATA(c_type)                                                                       \
+  case utils::ToTensorProtoElementType<c_type>(): {                                                         \
+    if (utils::HasType<EnabledOutputTypeList, c_type>()) {                                                  \
+      c_type val;                                                                                           \
+      auto unpack_status = utils::UnpackTensor(t_proto, raw_data, raw_data_len, &val, 1);                   \
+      ORT_ENFORCE(unpack_status.IsOK(), "Value attribute unpacking failed:", unpack_status.ErrorMessage()); \
+      SetValue(sizeof(c_type), reinterpret_cast<void*>(&val));                                              \
+    } else {                                                                                                \
+      ORT_THROW("Value attribute datatype is not enabled in this build: ", tensor_type);                    \
+    }                                                                                                       \
+    break;                                                                                                  \
+  }
+
+template <typename EnabledOutputTypeList>
+void ConstantOfShapeBase<EnabledOutputTypeList>::SetValueFromTensorProto(const ONNX_NAMESPACE::TensorProto& t_proto) {
+  ORT_ENFORCE(utils::HasDataType(t_proto));
+  ORT_ENFORCE(ONNX_NAMESPACE::TensorProto::DataType_IsValid(t_proto.data_type()));
+  ORT_ENFORCE(!utils::HasExternalData(t_proto),
+              "Tensor proto with external data for value attribute is not supported.");
+  const auto tensor_type = static_cast<ONNX_NAMESPACE::TensorProto_DataType>(t_proto.data_type());
+  const void* const raw_data = utils::HasRawData(t_proto) ? t_proto.raw_data().data() : nullptr;
+  const size_t raw_data_len = utils::HasRawData(t_proto) ? t_proto.raw_data().size() : 0;
+  switch (tensor_type) {
+    CASE_FETCH_VALUE_DATA(bool)
+    CASE_FETCH_VALUE_DATA(float)
+    CASE_FETCH_VALUE_DATA(MLFloat16)
+    CASE_FETCH_VALUE_DATA(double)
+    CASE_FETCH_VALUE_DATA(int8_t)
+    CASE_FETCH_VALUE_DATA(int16_t)
+    CASE_FETCH_VALUE_DATA(int32_t)
+    CASE_FETCH_VALUE_DATA(int64_t)
+    CASE_FETCH_VALUE_DATA(uint8_t)
+    CASE_FETCH_VALUE_DATA(uint16_t)
+    CASE_FETCH_VALUE_DATA(uint32_t)
+    CASE_FETCH_VALUE_DATA(uint64_t)
+    default:
+      ORT_THROW("Unsupported value attribute datatype: ", tensor_type);
+  }
+}
```

Comment:
Is it better to use a dispatcher here given otherwise you're defining code for all types even if they're disabled?

Reply:
tried using a dispatcher but wasn't able to match the binary size of the switch version when all types are enabled

---
In reply to: [573371802](https://github.com/microsoft/onnxruntime/pull/6594#discussion_r573371802) [](ancestors = 573371802)